modules/core/include/opencv2/core/quaternion.hpp

```diff
@@ -0,0 +1,1949 @@
+        {
+            CV_Error(Error::StsBadArg, "This quaternion does not represent a rotation");
+        }
+        return 2 * std::acos(w / norm());
+    }
+
+    template <typename T>
+    inline Vec<T, 3> Quat<T>::getAxis(bool assumeUnit) const
+    {
+        T angle = getAngle(assumeUnit);
+        if (assumeUnit)
+        {
+            return Vec<T, 3>{x, y, z} / std::sin(angle / 2);
+        }
+        return Vec<T, 3> {x, y, z} / (norm() * std::sin(angle / 2));
+    }
+
+    template <typename T>
+    cv::Mat Quat<T>::toRotMat4x4(bool assumeUnit) const
+    {
+        T a = w, b = x, c = y, d = z;
+        if (!assumeUnit)
+        {
+            Quat<T> qTemp = normalize();
+            a = qTemp.w;
+            b = qTemp.x;
+            c = qTemp.y;
+            d = qTemp.z;
+        }
+        cv::Matx<T, 4, 4> R{
+            1,                      0,                       0,                   0,
+            0,1 - 2 * (c * c + d * d), 2 * (b * c + a * d)    , 2 * (b * d - a * c),
+            0,2 * (b * c - a * d)    , 1 - 2 * (b * b + d * d), 2 * (c * d + a * b),
+            0,2 * (b * d + a * c)    , 2 * (c * d - a * b)    , 1 - 2 * (b * b + c * c)};
+        return cv::Mat(R).t();
+    }
+
+    template <typename T>
+    cv::Mat Quat<T>::toRotMat3x3(bool assumeUnit) const
+    {
+        T a = w, b = x, c = y, d = z;
+        if (!assumeUnit)
+        {
+            Quat<T> qTemp = normalize();
+            a = qTemp.w;
+            b = qTemp.x;
+            c = qTemp.y;
+            d = qTemp.z;
+        }
+        cv::Matx<T, 3, 3> R{
+            1 - 2 * (c * c + d * d), 2 * (b * c + a * d)    , 2 * (b * d - a * c),
+            2 * (b * c - a * d)    , 1 - 2 * (b * b + d * d), 2 * (c * d + a * b),
+            2 * (b * d + a * c)    , 2 * (c * d - a * b)    , 1 - 2 * (b * b + c * c)};
+        return cv::Mat(R).t();
+    }
+
+    template <typename T>
+    Vec<T, 3> Quat<T>::toRodrigues(bool assumeUnit) const
+    {
+        if (abs(w) < CV_QUAT_EPS)
+        {
+            CV_Error(Error::StsBadArg, "Cannot convert quaternion to rotation vector: rotation vector is indeterminte");
+        }
+        if (assumeUnit)
+        {
+            return Vec<T, 3>{x / w, y / w, z / w};
+        }
+        Quat<T> q = normalize();
+        return Vec<T, 3>{q.x / q.w, q.y / q.w, q.z / q.w};
+
+    }
+
+    template <typename T>
+    cv::Vec<T, 4> Quat<T>::toVec() const
+    {
+        return cv::Vec<T, 4>{w, x, y, z};
+    }
+
+    template <typename T>
+    Quat<T> Quat<T>::lerp(const Quat<T> &q0, const Quat<T> &q1, const T t)
+    {
+        return (1 - t) * q0 + t * q1;
+    }
+
+    template <typename T>
+    Quat<T> Quat<T>::slerp(const Quat<T> &q0, const Quat<T> &q1, const T t, bool assumeUnit, bool directChange)
+    {
+        Quatd v0(q0);
+        Quatd v1(q1);
+
+        if (!assumeUnit)
+        {
+            v0 = v0.normalize();
+            v1 = v1.normalize();
+            // add warning:
```

Comment:
There shouldn't be any warnings (about performance, I guess?), we do not guarantee good performance for quaternions which are not normalized.

Reply:
OK. This is commented before I write the function description earlier